lib/AST/DiagnosticEngine.cpp

```diff
@@ -850,27 +857,34 @@ DiagnosticBehavior DiagnosticState::determineBehavior(const Diagnostic &diag) {
 
   //   3) If the user provided a behavior for this diagnostic's kind, follow
   //      that
-  if (diagInfo.kind == DiagnosticKind::Warning) {
+  if (diagInfo.kind == DiagnosticKind::Warning
+      || (diag.getBehaviorLimit() == DiagnosticBehavior::Warning && !isNote)) {
```

Comment:
Is this going to end up suppressing/erroring if the diagnostic kind is a `Remark` and the behavior limit is `Warning`?

Reply:
You're right, this is wrong.

Reply:
The whole method was kind of confusing, so I've rewritten it so that we first determine the behavior that's being requested and then apply all of these state- and settings-related changes. It also now has more straightforward control flow, so hopefully this version will be better.

Reply:
Yes, it's definitely much easier to read now!